virttest/qemu_devices/qbuses.py

```diff
@@ -650,6 +650,118 @@ def _increment_addr(self, addr, last_addr=None):
+                        "virtio-scsi-pci", "virtio-balloon-pci",
+                        "virtio-serial-pci", "virtio-rng-pci",
+                        "e1000e", "virtio-gpu-device"]
+        if device.get_param("driver") in pcie_devices and (
+                    device.get_param("direct_plug_pci", "no") == "no"):
+            return False
+
+        return True
+
+    def add_root_port(self, addr, root_port=None, multifunction="off"):
+        """
+        Add pcie root port to the bus and __root_ports list, assign free slot,
+        chassis, and addr for it according to current list.
+
+        :param addr: addr to add the root port
+        :param root_port: pcie-root-port QDevice object
+        :param multifunction: if multifunction on or off
+        :return: pcie-root-port QDevice object
+        """
+        _addr = int(addr, 16)
+        if not root_port:
+            root_port = self._add_root_port(addr, multifunction)
+        if _addr not in self.__root_ports:
+            bus = root_port.child_bus[0]
+            params = {'slot': _addr, 'chassis': _addr, 'addr': "0x%x" % _addr}
+            for key, value in params.iteritems():
+                root_port.set_param(key, value)
+            self.__root_ports["0x%x" % _addr] = bus
+        return root_port
+
+    def _add_root_port(self, addr, multifunction="off"):
+        """
+        Generate pcie-root-port QDevice object with certain addr and id
+
+        :param addr: addr to add the root port
+        :param multifunction: if multifunction on or off
+        :return: pcie-root-port QDevice object
+        """
+        _addr = int(addr, 16)
+        if _addr not in self.__root_ports:
+            bus_id = "%s-root-port-%s" % (self.busid, _addr)
+            length = 1 if multifunction == "off" else 8
+            bus = QPCIBus(bus_id, 'PCIE', bus_id, length=length)
+            params = {'id': bus_id}
+            return qdevices.QDevice(self.__root_type,
+                                    params,
+                                    aobject=bus_id,
+                                    parent_bus=[{'aobject': self.aobject}],
+                                    child_bus=bus)
+
+    def _insert(self, device, addr):
+        """
+        Inserts a pcie-root-port before the device if it is not a
+        direct plug device.
+
+        :param device: qdevice.QDevice object
+        :param addr: addr to insert the device
+        """
+        _addr = addr.split('-')[0]
+        if self.is_direct_plug(device):
+            if device.get_param("driver") == self.__root_type:
+                device = self.add_root_port(_addr, device)
+            return super(QPCIEBus, self)._insert(device, addr)
+        added_devices = []
+        root_port = self.add_root_port(_addr)
+        if root_port is not None:
+            added_devices.append(root_port)
+            device['bus'] = root_port.child_bus[0].busid
+        return added_devices
```

Comment:
Well in this branch (when inserting a device) the `self.bus[addr]` is not updated, therefor the bus looks empty:

```
from virttest.qemu_devices.qbuses import *
>>> from virttest.qemu_devices.qdevices import *
>>> a=QPCIEBus("me", "PCI")
>>> a.insert(QDevice("virtio-scsi-pci"))
[<virttest.qemu_devices.qdevices.QDevice object at 0x7f7f4662ecd0>]
>>> a.insert(QDevice("virtio-scsi-pci"))
[<virttest.qemu_devices.qdevices.QDevice object at 0x7f7f4662ee10>]
>>> print a
me(PCI): {}
```

Reply:
OK, this happens on `insert` of the `pcie-root-port`